File: algorithms/4-MedianofTwoSortedArrays/solution.py
def get_max_from_arrays(arr1 ,index1,arr2 ,index2):
        if index1 == 0:
            return float(arr2[index2-1])
    
    
        if index2==0:
            return float(arr1[index1-1])
      
        return max(arr1[index1-1],arr2[index2-1])
    
def get_min_from_arrays(arr1,index1,arr2,index2):
    if index1 == len(arr1):
        return arr2[index2]
    if index2 == len(arr2):
        return arr1[index1]
    
    return min(arr1[index1],arr2[index2])
    
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        nums1Size=len(nums1)
        nums2Size=len(nums2)
        if nums1Size <= nums2Size:
            m,n = nums1Size,nums2Size
            A,B = nums1, nums2
        else:
            m,n = nums2Size,nums1Size
            A,B = nums2, nums1
        
        
        imin ,imax, half=0,m,int((m+n+1)/2)
        
        isOdd = (m+n) % 2 !=0
        
        while True :
            i = int((imin+imax)/2)
            j = half - i
            print(i,j)
            if (j==0 or i==m or B[j-1] <= A[i]) and (i==0 or j==n or A[i-1] <=B[j]):
                if isOdd:
                    print("odd:",i,j)
                    return get_max_from_arrays(A,i,B,j)
                else:
                    return (get_max_from_arrays(A,i,B,j)+ get_min_from_arrays(A,i,B,j))/2
                
            elif B[j-1] >A[i]:
                imin=i+1
                print(imin)
                
            elif A[i-1] >B[j]:
                imax=i-1
        
        return 0.0
    
    
    
        

File: algorithms/4-MedianofTwoSortedArrays/solution.java
class Solution {
    public double findMedianSortedArrays(int[] nums1, int[] nums2) {
        int num1size=nums1.length;
        int num2size=nums2.length;
        int[] A,B;
        int m,n;
        if(num1size <= num2size){
            m = num1size;
            n = num2size;
            A= nums1;
            B= nums2;
        }else{
            m = num2size;
            n = num1size;
            A= nums2;
            B= nums1;
        }
         
        // System.out.println(String.format("%s,%s",m,n));
        // System.out.println(String.format("%s,%s",A.length,B.length));
        int imin=0;
        int imax=m;
        int half=(m+n+1)/2;
        boolean isOdd =(m+n) %2 !=0;
        while (imin <= imax){
            int i= (imin+imax)/2;
            int j= half-i;
            
            if(i<m &&B[j-1]>A[i]){
                imin++;
            }else if(i>0 && A[i-1]>B[j]){
                imax--;
            }else{
                if(isOdd){
                    return getMaxLeft(A,B,i,j);
                }else{
                    return (getMaxLeft(A,B,i,j)+getMinRight(A,B,i,j))/2.0;
                }
            }
        }
        return 0;
    }
    
    public int getMaxLeft(int[] A,int[]B,int i,int j){
        if(i>0 &&j >0){
            return Math.max(A[i-1],B[j-1]);
        }else if(j>0){
            return B[j-1];
        }else {
            return A[i-1];
        }
    }
    
    public int getMinRight(int[] A,int[]B,int i,int j){
        if(i<A.length && j<B.length){
            return Math.min(A[i],B[j]);
        }else if(i < A.length){
            return A[i];
        }else{
            return B[j];
        }
    }
}
